### ISSUE TITLE
Improve Publish Masking + DataSubject Updates

### ISSUE THREAD

**Issue by madhavajay** [OpenMined/PySyft#7039]
## Description
Currently publish masking drops whole tensors if the entire tensor `max` epsilon is over the current budget.

This `zer0ing` should instead be applied as a mask of overbudget entities of the Tensor values and the removal of the DataSubjects at those positions. We probably need to remove the DataSubjects because if the left hand side scalar is zer0ed then 0 + 1 needs to only have the right hand DataSubject rather than a union of both sides. This is distinct from a true original `0` value so the filter process should do this when zeroing.

Current code:
```python
def filtered(self) -> GammaTensor:
    return GammaTensor(
        child=jnp.zeros_like(self.child),
        data_subjects=self.data_subjects,
        min_vals=self.min_vals * 0,
        max_vals=self.max_vals * 1,
        func_str=GAMMA_TENSOR_OP.NOOP.value,
    )

if epsilon > privacy_budget:
    filtered_tensor = input_tensor.filtered()
    parent_state[filtered_tensor.id] = filtered_tensor
```

To allow these changes to flow through the graph though we might need to change the data structure.

Currently, when a root PhiTensor is filtered it's values over the epsilon budget are zer0ed. 
This seems to directly update the children in the `jax` calculation graph which is cool. 

However in the current implementation the DataSubjectArrays (of sets) are new sets of the union of DataSubjects and do not react to this update. Perhaps if each successive operation tracked a list of the lists of Data Subjects that were passed in (or a list of the root sets), then during `zer0ing` the root datasubject set or list can be mutated and the down stream pointers should be updated. On the final level calculation of the final mask of ids might require some flattening of a potentially quite long list of lists, however the resultant structure should match the updated union of `Data Subjects` at each scalar position.

Imagine a flattened list of sets for the root data dimension of two different Tensors as `left_root` and `right_root`, these could be (2, 2) np.arrays for example:

```python
# datasubjects for each index 0, 1, 2, 3
left_root = [{1}, {1, 2}, {3}, {4}]
right_root = [{9}, {3}, {1}, {6, 7}]

# lets stack a few layers deep just to test out how we can flatten them correctly
# if its always a union of n inputs of the same shape (usually 2 but not always)
l1 = [left_root]
r1 = [right_root]
l2 = [l1, l1]
r2 = [r1, r1]
final = [l2, r2]
```

```python
> [[[[{1}, {1, 2}, {3}, {4}]], [[{1}, {1, 2}, {3}, {4}]]],
 [[[{9}, {3}, {1}, {6, 7}]], [[{9}, {3}, {1}, {6, 7}]]]]
```

```python
def sum_on_axis(a, b):
    if isinstance(a, set) or isinstance(b, set):
        # last level
        return set(a).union(b)
    new = []
    for i in range(len(a)):
        new.append(sum_on_axis(a[i], b[i]))
    if len(new) == 1:
        return new[0]
    if len(new) == 2:
        return sum_on_axis(new[0], new[1])
    return new
```

```python
result = sum_on_axis(l2, r2)
> [{1, 9}, {1, 2, 3}, {1, 3}, {4, 6, 7}]
```

The above code is probably wrong so please check it thoroughly. Its likely we need augment the code to handle more than 2 inputs at a level, so probably that last bit of code should be something like if len(new) > 1, then iterate through and reduce that level?

## Acceptance Criteria
- A test which demonstrates that when the masked values are filtered the updated resultant Tensor calculates the correct final mask of DataSubjects per Scalar position and collects their RDP Constants or Epsilons for accounting with the `DataSubjectLedger`.
- The test includes a variety of ops such as single `self` ops, `infix` ops, `ops` with more than 2 inputs, `ops` which change the shape.

**Comment by madhavajay** [comment]
When planning the new DataSubjectArray data structure we need to keep in mind the need to differentiate between `homogenous` and `heterogeneous` scalar owners.

https://github.com/OpenMined/PySyft/issues/6934

**Comment by iamtrask** [comment]
Please schedule a design meeting with me to discuss the design of DataSubjectArray

**Comment by teo-milea** [comment]
We have agreed to move this on `0.8` and for the moment only filter entire tensors.

**Comment by madhavajay** [comment]
@teo-milea What about the Data Subjects for 0.7?

**Comment by madhavajay** [comment]
The data subjects work will be done in this other issue: https://github.com/OpenMined/PySyft/issues/7142

This issue is now specifically related to the advanced filtering of Tensors.
